engine/src/main/java/io/zeebe/engine/processing/streamprocessor/ReProcessingStateMachine.java

```diff
@@ -164,34 +165,42 @@ public ReProcessingStateMachine(final ProcessingContext context) {
    *
    * @return a ActorFuture with last reprocessed position
    */
-  ActorFuture<Long> startRecover(final long snapshotPosition) {
+  ActorFuture<Positions> startRecover(final long snapshotPosition) {
     recoveryFuture = new CompletableActorFuture<>();
-
     this.snapshotPosition = snapshotPosition;
 
-    LOG.trace("Start scanning the log for error events.");
-    lastSourceEventPosition = scanLog(snapshotPosition);
-    LOG.trace("Finished scanning the log for error events.");
+    LOG.trace("Start scanning the log for error events and last positions.");
+    positions = restoreProcessingState(snapshotPosition);
+    LOG.trace("Finished scanning the log for error events. Restored positions {}.", positions);
 
-    if (lastSourceEventPosition > snapshotPosition) {
+    if (positions.getLastSuccessfulProcessedPosition() > snapshotPosition) {
       LOG.info(
           "Processor starts reprocessing, until last source event position {}",
-          lastSourceEventPosition);
+          positions.getLastSuccessfulProcessedPosition());
       logStreamReader.seekToNextEvent(snapshotPosition);
       reprocessNextEvent();
     } else if (snapshotPosition > 0) {
-      recoveryFuture.complete(snapshotPosition);
+      positions.setLastSuccessfulProcessedPosition(snapshotPosition);
+      recoveryFuture.complete(positions);
     } else {
-      recoveryFuture.complete(StreamProcessor.UNSET_POSITION);
+      recoveryFuture.complete(positions);
```

Comment:
Maybe pass a copy of positions along or make `Positions` immutable.

Reply:
Why do you think this would be necessary?

Reply:
Not necessary, but safer.

This would prevent that the callee modifies the position while being called.